File: ios-app/Targets/Repository/Sources/ContactsRepository.swift
//
//  ContactsRepository.swift
//  Contacts
//
//  Created by Răzvan Roşu on 13/02/24.
//  Copyright © 2024 tuist.io. All rights reserved.
//

public final class ContactsRepository {
    private var sections: [ContactId : Contact] = mockContacts
        .map { contact in
            (contact.id, contact)
        }
        .reduce(into: [:]) { dictionary, tuple in
            dictionary[tuple.0] = tuple.1
        }
    
    public init() { }
    
    public func add(new contact: NewContact) {
        let id = ContactId()
        sections[id] = Contact(id: id, name: contact.name)
    }
    
    public func update(contact: Contact) {
        sections[contact.id] = contact
    }
    
    public func remove(id: ContactId) {
        sections.removeValue(forKey: id)
    }
    
    public func get() -> [ContactsSection] {
        Dictionary(
            grouping: sections.values.sorted(by: { $0.name < $1.name }),
            by: { $0.name.first! }
        )
            .map { initial, contacts in
                ContactsSection(initial: initial, contacts: contacts)
            }
            .sorted(by: { $0.initial < $1.initial })
    }
    
    public func get(id: ContactId) -> Contact {
        sections[id]!
    }
}

File: repository/src/main/kotlin/red/razvan/hello/kotlinmultiplatform/contacts/repository/ContactsRepository.kt
package red.razvan.hello.kotlinmultiplatform.contacts.repository

class ContactsRepository {
    private val contacts: MutableMap<ContactId, Contact> = mockContacts
        .associateBy(Contact::id)
        .toMutableMap()

    fun add(newContact: NewContact) {
        val id = ContactId()
        contacts[id] = Contact(id = id, name = newContact.name)
    }

    fun removeById(id: ContactId) {
        contacts.remove(id)
    }

    fun update(contact: Contact) {
        contacts[contact.id] = contact
    }

    fun getById(id: ContactId): Contact = requireNotNull(contacts[id])

    fun getAll(): List<ContactsSection> =
        contacts
            .values
            .asSequence()
            .sortedBy(Contact::name)
            .groupBy { contact ->
                contact.name.first()
            }
            .map {
                ContactsSection(initial = it.key, contacts = it.value)
            }
            .sortedBy(ContactsSection::initial)
}